### PR TITLE
Consumable items [WIP]

### PR DESCRIPTION
**Pull request by Wouterz90**
AC is seems to be fully consumable now, we can't have the icon for it showing up since it's tied to the ability (the container). Needs a cool icon for that.

### CODE REVIEW

**Comment by darklordabc** [comment]
Got this error when I tried to consume the AC: 
```
[   Developer            ]: AddAbility - Failed to create ability ability_consumable_item_container.
[   Developer            ]: ...dons\tester\scripts\vscripts\abilities\items\assault.lua:20: attempt to index local 'ab' (a nil value)
[   Developer            ]: stack traceback:
[   Developer            ]: 	...dons\tester\scripts\vscripts\abilities\items\assault.lua:20: in function 'ConsumeItem'
[   Developer            ]: 	...dons\tester\scripts\vscripts\abilities\items\assault.lua:10: in function <...dons\tester\scripts\vscripts\abilities\items\assault.lua:7>
[ W VScript              ]: Script Runtime Error: ...dons\tester\scripts\vscripts\abilities\items\assault.lua:20: attempt to index local 'ab' (a nil value)
[ W VScript              ]: stack traceback:
[ W VScript              ]: 	...dons\tester\scripts\vscripts\abilities\items\assault.lua:20: in function 'ConsumeItem'
[ W VScript              ]: 	...dons\tester\scripts\vscripts\abilities\items\assault.lua:10: in function <...dons\tester\scripts\vscripts\abilities\items\assault.lua:7>
```

**Comment by Wouterz90** [comment]
Might have been no space for the ability, but that should have been fixed as well

No it was missing a .txt in the base file

Oh and please do not change those variable names, since we need to have them in the ability it really needs the item name in front of it.

**Comment by darklordabc** [comment]
Still need to add texts to all the modifiers, also the Assault cuirass has two modifiers for the hosts, one should probably be invisible. You should not be able to consume an object if you have already consumed one of its type earlier, it doesnt give any benefit to consume twice so players shouldnt be allowed to do it.

**Comment by Wouterz90** [comment]
I was thinking about doing all the tooltips etc in the end, since it's annoying to recompile everything over and over.

**Comment by darklordabc** [comment]
So is this finished?

**Comment by Wouterz90** [comment]
If you dont find any unwanted behavior, it is.

**Comment by darklordabc** [comment]
When I bought AC (not consume just bought), I got this error. After the error I was unable to consume any item.
```
[   General              ]: **** Panel creditsDisplay has fill-parent-flow for width, but isn't in a flowing right layout
[   Developer            ]: ...dons\tester\scripts\vscripts\abilities\items\assault.lua:50: attempt to index global 'modifier_item_desolator_consumable' (a nil value)
[   Developer            ]: stack traceback:
[   Developer            ]: 	...dons\tester\scripts\vscripts\abilities\items\assault.lua:50: in main chunk
[   Developer            ]: 	[C]: in function 'AddItemByName'
[   Developer            ]: 	scripts/vscripts/ingame.lua:151: in function <scripts/vscripts/ingame.lua:144>
[ W VScript              ]: Script Runtime Error: ...dons\tester\scripts\vscripts\abilities\items\assault.lua:50: attempt to index global 'modifier_item_desolator_consumable' (a nil value)
[ W VScript              ]: stack traceback:
[ W VScript              ]: 	...dons\tester\scripts\vscripts\abilities\items\assault.lua:50: in main chunk
[ W VScript              ]: 	[C]: in function 'AddItemByName'
[ W VScript              ]: 	scripts/vscripts/ingame.lua:151: in function <scripts/vscripts/ingame.lua:144>
[ W VScript              ]: Error running script named abilities/items/assault.lua
```

This is probably related to the above error, but when I bought a desolator, it was immediately deleted from my inventory. 
Alt Text

**Comment by darklordabc** [comment]
When you get a consumable item, it automatically moves to your inventory even if your not near a shop, it shouldn't do this. Also, when I buy a deadalous, it immeditedly delete itself from my inventory. 
https://i.gyazo.com/064e430907769395745c4c3d396590c6.gif

**Comment by darklordabc** [comment]
When I consume the Skadi (for example) I gain the modifier buff as I should, BUT, when I buy another Skadi, I immediately lose that modifier icon (and probably the modifer itself), and If I drop the second Skadi I bought, I no longer have any benefit from the consumed one.

**Comment by Wouterz90** [comment]
Somehow I didn't add that for skadi, for the other items it was already working. I did not have your deadalus issue. Items will now get changed when they are in your inventory.

**Comment by darklordabc** [comment]
- [ ] I buy the skadi away from base and the courier delievers it but it doesnt turn into a consumable when I get it
image

- [ ] Daedalus straight vanishes as soon as I buy it. Same as heart. Console says this:
[ W General ]: Could not find data for item! item_greater_crit_consumable
[ W General ]: Could not find data for item! item_greater_crit_consumable
[ W General ]: Could not find data for item! item_heart_consumable
[ W General ]: Could not find data for item! item_heart_consumable

- [ ] Non-consumed version of the consumable assault cuirass, does not give benefits, I'm worried that some other consumable items may not be giving benefits either. First pic is just holding the AC, the second is after consuming it.
image

**Comment by Wouterz90** [comment]
Hmm looks like I have work to do, will do it tomorrow after work.

For the items it might be that your local items file isn't the one in this repo, else no idea because it works for me.

**Comment by Wouterz90** [comment]
There is an issue when items are partially formed from the courier, should use another event to detect things like those I guess. Dropping and picking the items up could be a small fix for now.

**Comment by Wouterz90** [comment]
I have tested all the events, there isn't one that does catch an inventory change through courier so we might need to periodically check for them. Dropping as an option will mean that the old items can be kept which would allow stacking.

**Comment by Wouterz90** [comment]
I can't find any issues with this anymore

**Comment by darklordabc** [comment]
I'll test soon

**Comment by Wouterz90** [comment]
assault_ was missing in the items file for the assault cuirrass, that's why it didn't do anything.

I removed all the item tags because I did not want them to be visible in the shop, because players are buying the regular items. Announcing them means it will announce a skadi twice.

**Comment by darklordabc** [comment]
I can still buy the assault cuirass from the shop, and it doesnt auto convert into a consumerable version.

File: src/game/scripts/vscripts/ingame.lua [hunk 1]
```diff
@@ -153,6 +153,33 @@ function Ingame:OnHeroLeveledUp(keys)
     -- print(hero:GetUnitName(), level, hero:GetDeathXP(), GetXPForLevel( level ))
 end
 
+function Ingame:FilterItemsInventory(keys)
```
Comment (darklordabc) [inline src/game/scripts/vscripts/ingame.lua]: Can we remove this if we dont need it.

File: src/game/scripts/vscripts/ingame.lua [hunk 2]
```diff
@@ -493,6 +523,7 @@ function Ingame:onStart()
     GameRules:GetGameModeEntity():SetTrackingProjectileFilter(Dynamic_Wrap(Ingame, 'FilterProjectiles'), self)
     GameRules:GetGameModeEntity():SetModifierGainedFilter(Dynamic_Wrap(Ingame, 'FilterModifiers'),self)  
     GameRules:GetGameModeEntity():SetDamageFilter(Dynamic_Wrap(Ingame, 'FilterDamage'),self)
+    GameRules:GetGameModeEntity():SetItemAddedToInventoryFilter(Dynamic_Wrap(Ingame, 'FilterItemsInventory'),self)
```
Comment (darklordabc) [inline src/game/scripts/vscripts/ingame.lua]: Can we remove this if we dont need it.

File: src/game/scripts/vscripts/ingame.lua [hunk 3]
```diff
@@ -541,6 +572,31 @@ function Ingame:fixRuneBug()
...
+            local hero = PlayerResource:GetSelectedHeroEntity(i)
+            if hero and IsValidEntity(hero) then
+                for i=0,14 do
+                    local hItem = hero:GetItemInSlot(i)
+                    if hItem then
+                        local name = hItem:GetAbilityName()
+                        if itemTable[name] then
+                            hero:RemoveItem(hItem)
+                            hero:AddItemByName(name.."_consumable")
+                            local nSlot, hUseless = hero:FindItemByNameEverywhere(name.."_consumable")
```
Comment (darklordabc) [inline src/game/scripts/vscripts/ingame.lua:566]: What is the "hUseless" variable used for?

Comment (Wouterz90) [reply src/game/scripts/vscripts/ingame.lua:566]: It's the consumable item, it's not used, but the function returns it anyways so I thought I'd keep it around for future problems.

**Comment by darklordabc** [comment]
Unlike all the other consumables, Heart shows its buff before you consume it, while its in your inventory, it makes the tooltip description a bit akward, can you make it hidden until you consume, like all the other items. 
image
When the aura returns from de-activation, its missing its green ring
image
Roshan does not de-activate the buff. Heroe damage (excluding some like heart-stopper aura) and Roshan are the only things that are meant to be able to break the regen. 
image
This is a problem with the gold cost. Players can attempt a consume even when they have the modifier, nothing will happen because they can only have 1 , but the gold will be consumed. In the ability could you make it that it does GetGoldCost of the ability and refunds that much money back to the player.

**Pull request merged**